File: Session12/784. Letter Case Permutation.py
class Solution:
    def letterCasePermutation(self, s: str) -> List[str]:
        self.s=s
        n=len(s)
        self.ans=[]
        self.helper(0,n)
        return self.ans
    
    def helper(self,pos,n):
        if pos>=n:
            self.ans.append(self.s)
            return
        
        if self.s[pos].isalpha():
            self.s=self.s[0:pos]+self.s[pos].upper()+self.s[pos+1:]
            self.helper(pos+1,n)
            self.s=self.s[0:pos]+self.s[pos].lower()+self.s[pos+1:]
            self.helper(pos+1,n)
        else:
            self.helper(pos+1,n)

File: Session12/784. Letter Case Permutation.java
class Solution {
    ArrayList<String> ans=new ArrayList<>();
    public List<String> letterCasePermutation(String s) {
        backtrack(s.toCharArray(),s.length(),0);
        return ans;
    }
    
    public void backtrack(char arr[],int n,int start)
    {
        if(start>=n)
        {
            ans.add(String.valueOf(arr));
            return;
        }
        if(Character.isAlphabetic(arr[start]))
        {
            arr[start]=Character.toLowerCase(arr[start]);
            backtrack(arr,n,start+1);
            arr[start]=Character.toUpperCase(arr[start]);
            backtrack(arr,n,start+1);
        }
        else
        {
            start++;
            backtrack(arr,n,start);
        }
    }
}